yarn-project/simulator/src/avm/journal/journal.ts

```diff
@@ -92,33 +100,42 @@ export class AvmPersistableStateManager {
     this.log.debug(`Storage write (address=${contractAddress}, slot=${slot}): value=${value}`);
     // Cache storage writes for later reference/reads
     this.publicStorage.write(contractAddress, slot, value);
-    const writeTreeSlot = computePublicDataTreeLeafSlot(contractAddress, slot);
-    const result = await this.merkleTree.batchInsert(
-      MerkleTreeId.PUBLIC_DATA_TREE,
-      [new PublicDataTreeLeaf(writeTreeSlot, value).toBuffer()],
-      0,
-    );
-    this.log.verbose(`Inserted public data tree leaf at slot ${slot} with value: ${value}`);
-    const lowLeafInfo = result.lowLeavesWitnessData![0];
-    const insertionPath = result.newSubtreeSiblingPath.toFields();
-    const lowLeafPreimage = PublicDataTreeLeafPreimage.fromBuffer(lowLeafInfo.leafPreimage.toBuffer());
-    const lowLeafIndex = lowLeafInfo.index;
-    const lowLeafPath = lowLeafInfo.siblingPath.toFields();
-    const newLeafPreimage = new PublicDataTreeLeafPreimage(
-      writeTreeSlot,
-      value,
-      lowLeafPreimage.nextSlot,
-      lowLeafPreimage.nextIndex,
-    );
-    this.trace.tracePublicStorageWrite(
-      contractAddress,
-      slot,
-      lowLeafPreimage,
-      new Fr(lowLeafIndex),
-      lowLeafPath,
-      newLeafPreimage,
-      insertionPath,
-    );
+    const leafSlot = computePublicDataTreeLeafSlot(contractAddress, slot);
+    if (this.doMerkleOperations) {
+      const result = await this.merkleTrees.batchInsert(
+        MerkleTreeId.PUBLIC_DATA_TREE,
+        [new PublicDataTreeLeaf(leafSlot, value).toBuffer()],
+        0,
+      );
+      assert(result !== undefined, 'Public data tree insertion error. You might want to disable skipMerkleOperations.');
+      this.log.debug(`Inserted public data tree leaf at leafSlot ${leafSlot}, value: ${value}`);
+
+      const lowLeafInfo = result.lowLeavesWitnessData![0];
+      const lowLeafPreimage = lowLeafInfo.leafPreimage as PublicDataTreeLeafPreimage;
+      const lowLeafIndex = lowLeafInfo.index;
+      const lowLeafPath = lowLeafInfo.siblingPath.toFields();
+
+      const insertionPath = result.newSubtreeSiblingPath.toFields();
+      const newLeafPreimage = new PublicDataTreeLeafPreimage(
+        leafSlot,
+        value,
+        lowLeafPreimage.nextSlot,
+        lowLeafPreimage.nextIndex,
+      );
+      // FIXME: Why do we need to hint both preimages for public data writes, but not for nullifier insertions?
+      this.trace.tracePublicStorageWrite(
+        contractAddress,
+        slot,
+        value,
```

Comment:
added back in `value` because it makes it easier to test and makes it usable while completely omitting the merkle args (when `doMerkleOperations = false`